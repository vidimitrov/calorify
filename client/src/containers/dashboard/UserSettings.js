import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from '@reach/router';
import { isEmpty } from 'lodash';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import ArrowBack from '@material-ui/icons/ArrowBack';
import Button from '@material-ui/core/Button';
import Form from '../../components/dashboard/Form';
import FormControls from '../../components/dashboard/FormControls';
import InputWrapper from '../../components/dashboard/InputWrapper';
import Logo from '../../components/dashboard/Logo';
import Wrapper from '../../components/dashboard/Wrapper';
import CustomTypography from '../../components/dashboard/CustomTypography';
import CustomIconButton from '../../components/dashboard/CustomIconButton';
import CustomTextField from '../../components/dashboard/CustomTextField';
import logo from '../../assets/img/logo.png';
import { updateUser as updateUserActionCreator } from '../../actions/users';

export class UserSettings extends React.Component {
  constructor(props) {
    super(props);

    const { user } = props;
    const { name, dailyCaloriesLimit } = user || {};

    this.state = {
      name,
      dailyCalories: dailyCaloriesLimit,
      change: false,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
  }

  onChangeHandler(key, value) {
    this.setState({
      change: true,
      [key]: value,
    });
  }

  render() {
    const {
      user,
      updateUser,
    } = this.props;
    const {
      name,
      dailyCalories,
      change,
    } = this.state;
    const { navigate } = this.props;

    if (!user) {
      return (
        <Redirect to="/auth/login" />
      );
    }

    return (
      <Wrapper>
        <AppBar position="static">
          <Toolbar>
            <CustomIconButton color="secondary" aria-label="Menu">
              <ArrowBack onClick={() => navigate('/')} />
            </CustomIconButton>
            <Logo src={logo} alt="logo" />
            <CustomTypography variant="h6" color="secondary">
              User Settings
            </CustomTypography>
          </Toolbar>
        </AppBar>
        <Form>
          <FormControls>
            <InputWrapper>
              <CustomTextField
                label="Full name"
                value={name}
                id="name"
                name="name"
                onChange={e => this.onChangeHandler('name', e.target.value)}
              />
            </InputWrapper>
            <InputWrapper>
              <CustomTextField
                label="Daily calories"
                helperText="It's your expected calorie intake per day"
                value={dailyCalories}
                id="calories"
                name="calories"
                onChange={e => this.onChangeHandler('dailyCalories', e.target.value)}
              />
            </InputWrapper>
          </FormControls>
          {/* TODO: It should be disabled if no changes are made */}
          <Button
            variant="contained"
            color="primary"
            name="submit-account-update"
            disabled={!change}
            onClick={() => {
              updateUser(user.id, {
                name,
                dailyCalories,
              }).then(() => {
                // TODO: Show positive snackbar
              }).catch(() => {
                // TODO: Show negative snackbar. Try again
              });
            }}
          >
            Save Changes
          </Button>
        </Form>
      </Wrapper>
    );
  }
}

UserSettings.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    role: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    dailyCaloriesLimit: PropTypes.number.isRequired,
  }),
  updateUser: PropTypes.func.isRequired,
  navigate: PropTypes.func.isRequired,
};

UserSettings.defaultProps = {
  user: null,
};

const mapStateToProps = state => ({
  user: isEmpty(state.user.data) ? null : state.user.data,
});

const mapDispatchToProps = dispatch => ({
  updateUser: (userId, updates) => dispatch(updateUserActionCreator(userId, updates)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserSettings);
