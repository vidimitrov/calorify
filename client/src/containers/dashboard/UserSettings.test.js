import React from 'react';
import renderer from 'react-test-renderer';
import { UserSettings } from './UserSettings';

it('UserSettings container renders correctly', () => {
  const props = {
    updateUser: () => { },
    navigate: () => { },
    user: {
      id: 'some-fake-id',
      name: 'John',
      dailyCaloriesLimit: 2000,
      role: 'user',
      email: 'john@email.com',
    },
  };
  const tree = renderer.create(
    <UserSettings {...props} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
