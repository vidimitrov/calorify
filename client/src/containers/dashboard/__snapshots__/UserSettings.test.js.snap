// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings container renders correctly 1`] = `
<div
  className="Wrapper-m60uq5-0 hkWJUG"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-47 MuiIconButton-root-41 MuiIconButton-colorSecondary-44 CustomIconButton-sc-1edavx9-0 eiNqgO"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-46"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-50"
            focusable="false"
            onClick={[Function]}
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-187"
        />
      </button>
      <img
        alt="logo"
        className="Logo-sc-1vcm3wd-0 lmrySJ"
        src="logo.png"
      />
      <h6
        className="MuiTypography-root-59 MuiTypography-h6-76 MuiTypography-colorSecondary-90 CustomTypography-e31nmz-0 kgGAHm"
      >
        User Settings
      </h6>
    </div>
  </header>
  <div
    className="Form-ycfidt-0 jwaFmv"
  >
    <div
      className="FormControls-sc-7fqkjw-0 dcNCfX"
    >
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <label
            className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
            data-shrink={true}
            htmlFor="name"
          >
            Full name
          </label>
          <div
            className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-139 MuiInput-input-124"
              disabled={false}
              id="name"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="John"
            />
          </div>
        </div>
      </div>
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          aria-describedby="calories-helper-text"
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <label
            className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
            data-shrink={true}
            htmlFor="calories"
          >
            Daily calories
          </label>
          <div
            className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-139 MuiInput-input-124"
              disabled={false}
              id="calories"
              name="calories"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value={2000}
            />
          </div>
          <p
            className="MuiFormHelperText-root-146 MuiFormHelperText-filled-152"
            id="calories-helper-text"
          >
            It's your expected calorie intake per day
          </p>
        </div>
      </div>
    </div>
    <button
      className="MuiButtonBase-root-47 MuiButtonBase-disabled-48 MuiButton-root-154 MuiButton-contained-165 MuiButton-containedPrimary-166 MuiButton-raised-168 MuiButton-raisedPrimary-169 MuiButton-disabled-174"
      disabled={true}
      name="submit-account-update"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-155"
      >
        Save Changes
      </span>
    </button>
  </div>
</div>
`;
